news/_posts/2019-10-10-about-overriding-system.md

```diff
@@ -0,0 +1,131 @@
+
+We provide native (or Core) classes in PrestaShop, named *CustomerForm* and *FormFormatter* but our implementation doesn't fits his client needs. So his strategy was to override the 2 classes and use their own, extending the core class to not have to rewrite everything.
+
+We didn't discuss details that much the developers needs, but we can imagine he wants to be able to override this function in charge of defining the Customer form:
+
+```php
+<?php
+// override/classes/form/CustomerFormatter
+use CustomerFormatter as NativeCustomerFormatter;
+
+class CustomerFormatter extends NativeCustomerFormatter
+{
+    public function getFormat()
+    {
+        // Here is the business logic to create the form
+        // with the right fields.
+    }
+}
+```
+
+Doing that allow this class to be used as a replacement of the original class in the Core.
+
+But... the core implementation is not "extensible" enough: this is a bug! Most of its properties are private and a function named `addConstraints` (in charge of  the Form validation) is private!
+
+Also, he'd like to use the translator directly from the customer.
+
+His request was reasonnable, but we won't be able to accept it.
+
+### The end of the overriding system
+
+Overrides are here since 9 years, and we now have a lot of feedbacks about it. Because developers are allowed to alter behavior of Core from their modules and because we are unable to ensure that Core update won't break updates from their modules, we didn't have fixed the "upgrade terrible mistake" issue, we only have taken the responsibility of it.
+
+The truth is that the only person able to ensure that its override won't break when the core will be upgraded is the person in charge of the override. In nine years, upgrading a shop is still a terrible experience and most of the times people don't upgrade their shop and keep their business outdated (technical point of view). We have worked hard on the upgraded module, describing more and more information about what could go wrong during the upgrade: databases migrations, non existing/deprecated classes functions/cache issues and so on.
+
+> We can say that undoubtely, Overriding system was a bad idea and we
+> are deprecating it.
+
+### Overriding system deprecation in 1.7.x releases
+
+As most of you have noticed, it's not possible to override core classes located in the `src` folder: this is intended and not something we will implement/introduce back later.
+
+The override system is still available on "soon-to-be-migrated" pages, but we strongly discourage you to rely on it as when pages will be migrated, this feature won't be backported.
+
+The override system is deprecated in PrestaShop Front Office (and on the Customer Back Office): this will works until the end of 1.7.x and we will probably add some warnings when used to help developers to prepare the migration to the next major version.
+
```

Comment:
As they are using PSR-2, they could be autoloaded by Composer instead of the PrestaShop autoloader.

If we don't have time to rename them accordingly, we just have to remove the "overriding feature" part from the PrestaShop [autoloader](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/PrestaShopAutoload.php#L135) so only native classes will be found.